### ISSUE TITLE
Make empty braces optional

### ISSUE THREAD

**Issue by romirk** [Mango0x45/gsp#1]
```
meta charset="UTF-8"
```

should be valid syntax

**Comment by Mango0x45** [comment]
```
div .class-1
div .class-2
```

Which of the following should that compile to?

```
<div class="class-1">
<div class="class-2">
```

```
<div class="class-1 class-2" div>
```

The braces solve ambiguities.

**Comment by romirk** [comment]
Should compile into empty tags:

```
<div class="class-1"></div>
<div class="class-2"></div>
```

**Comment by Mango0x45** [comment]
But what if I have a really long list of attributes?  Putting them on multiple lines might be useful for readability:

```
div
  #id
  .class-1
  .class-2
  attr-1="foo"
  attr-2="bar"
{}
```

We _could_ use indentation-based scoping here to figure out that the attributes all belong to the `div`, but that would that you can no longer use macro processors with the language as easily, as they would need to know the indentation level of the markup they generate in advance.

**Comment by Mango0x45** [comment]
> Should compile into empty tags:

Also, why is that?  If you want empty tags, you can simply use an empty text node via `node{-}`.  It feels weird to have the default be `<tag></tag>` instead of simply `<tag>` (a self-closing HTML tag).  Meta-tags especially in HTML are never written with a closing tag, since they’re self-closing.

**Comment by romirk** [comment]
> feels weird to have the default be `<tag></tag>` instead of simply `<tag>` (a self-closing HTML tag). 

You're right, and I agree, tags that can self-close should self-close by default. 

I also agree with you about indentation. I have two proposed solutions: 

- allow optional brackets (any kind you prefer) around the attributes:

```
img [
    #img1
    src=foo
    bar=baz
]
``` 

- if there has to be something at the end of the list of attributes to disambiguate the start of the next tag, then maybe something like `/` to denote an empty/self-closing tag would work:

```
meta charset=utf-8 viewport=... /
```

**Comment by Mango0x45** [comment]
> if there has to be something at the end of the list of attributes to disambiguate the start of the next tag, then maybe something like / to denote an empty/self-closing tag would work:

I mean it would work, but it only saves 1 character compared to `meta … {}`, and adds (slight) complexity to the parser.

> allow optional brackets (any kind you prefer) around the attributes:
Not sure how I feel about this.  I need to think about it

**Comment by romirk** [comment]
Wait a second, you always know when something is or isn't an attribute because attributes are `([\.#]\w+)|(\w+=\w*)` whereas tags do not contain a period, hash  or equals sign within their names. You always unambiguously know when a tag and its associated list of attributes end.

**Comment by Mango0x45** [comment]
Not quite.  Attributes don’t need to have values attached to them:

    form { button checked {-I am checked!} }
